src/solc_0.5/Land/erc721/LandBaseTokenV3.sol

```diff
@@ -286,231 +405,247 @@ contract LandBaseTokenV3 is ERC721BaseTokenV2 {
+                    }
+                }
+            }
+        }
+
+        quadCompareSize = quadCompareSize / 2;
+        if (quadCompareSize >= 3)
+            (index, landMinted) = _checkAndClearOwner(land, quadMinted, landMinted, index, quadCompareSize);
+        return (index, landMinted);
     }
 
-    function _regroup3x3(address from, address to, uint256 x, uint256 y, bool set) internal returns (bool) {
-        uint256 id = x + y * GRID_SIZE;
-        uint256 quadId = LAYER_3x3 + id;
+    function _regroupQuad(
+        address from,
+        address to,
+        Land memory land,
+        bool set,
+        uint256 childQuadSize
+    ) internal returns (bool) {
+        (uint256 layer, , uint256 childLayer) = _getQuadLayer(land.size);
+        uint256 quadId = _getQuadId(layer, land.x, land.y);
         bool ownerOfAll = true;
-        for (uint256 xi = x; xi < x+3; xi++) {
-            for (uint256 yi = y; yi < y+3; yi++) {
-                ownerOfAll = _checkAndClear(from, xi + yi * GRID_SIZE) && ownerOfAll;
+
+        {
+            for (uint256 xi = land.x; xi < land.x + land.size; xi += childQuadSize) {
+                for (uint256 yi = land.y; yi < land.y + land.size; yi += childQuadSize) {
+                    uint256 ownerChild;
+                    bool ownAllIndividual;
+                    if (childQuadSize < 3) {
+                        ownAllIndividual = _checkAndClear(from, _getQuadId(LAYER_1x1, xi, yi)) && ownerOfAll;
+                    } else {
+                        ownAllIndividual = _regroupQuad(
+                            from,
+                            to,
+                            Land({x: xi, y: yi, size: childQuadSize}),
+                            false,
+                            childQuadSize / 2
+                        );
+                        uint256 idChild = _getQuadId(childLayer, xi, yi);
+                        ownerChild = _owners[idChild];
+                        if (ownerChild != 0) {
+                            if (!ownAllIndividual) {
+                                require(ownerChild == uint256(from), "not owner of child Quad");
+                            }
+                            _owners[idChild] = 0;
+                        }
+                    }
+                    ownerOfAll = (ownAllIndividual || ownerChild != 0) && ownerOfAll;
+                }
             }
         }
-        if(set) {
-            if(!ownerOfAll) {
-                require(
-                    _ownerOfQuad(3, x, y) == from,
-                    "not owner of all sub quads nor parent quads"
-                );
+
+        if (set) {
+            if (!ownerOfAll) {
+                require(_ownerOfQuad(land.size, land.x, land.y) == from, "not owner of all sub quads nor parent quads");
             }
             _owners[quadId] = uint256(to);
             return true;
         }
+
         return ownerOfAll;
     }
```

Comment:
I think the regroupQuad function is not easy to undeerstand & deserves some comments of the logic